src/libsyntax/parse/lexer/tokentrees.rs

```diff
@@ -204,10 +204,10 @@ impl<'a> TokenTreesReader<'a> {
                 Err(err)
             },
             _ => {
-                let tt = TokenTree::Token(self.token.take());
+                let token = self.token.take();
                 self.real_token();
-                let is_joint = self.joint_to_prev == Joint && self.token.is_op();
-                Ok((tt, if is_joint { Joint } else { NonJoint }))
+                let is_joint = self.joint_to_prev == Joint && token.is_op() && self.token.is_op();
```

Comment:
Perhaps it's time to move the `is_op` check to the proc macro API boundary (and track precise jointness for everything internally)?

Reply:
Makes sense. I've modified this line to avoid modifying a unit test, but I don't see why can't we move it entirely to proc-macro layer. (although I'd rather do it in a separate PR).